[Core] Allow vLLM to stream n tokens at a time

rohingarg-c opened the pull request:
This PR allows vLLM to stream n tokens at a time, improving potential network bandwidth issues. It stores this stream_n param as a field in a new StreamingParams class, created to store other parameters related to how the model streams tokens back.

rohingarg-c committed:
feat: add stream_n parameter support

Signed-off-by: Rohin Garg <rohin@character.ai>

yinghai commented on vllm/engine/async_llm_engine.py:
```diff
@@ -1068,7 +1072,19 @@ async def generate(
+
+                buffer_token_count += sum(
+                    len(o.token_ids) for o in output.outputs)
+                if buffer_token_count >= streaming_params.stream_n:
+                    yield buffer
+                    buffer = None
+                    buffer_token_count = 0
```
Seems legit. This is v0 engine? Maybe we don't need to support it?

hypnopump commented:
Like this feature

njhill reviewed (commented):
Thanks @character-tech. Like @yinghai suggested, I don't think we need to add this for the v0 path which is in feature-freeze right now (async_llm_engine.py, client.py).

For naming of the streaming param would something like `chunk_size` make more sense?

I'm also wondering whether this should be configurable globally rather than (or perhaps in addition to) per-request.

Could you improve the test to actually check the behaviour in terms of number of tokens per response message?

rohingarg-c committed:
removed changes from v0 engine, used RequestOutputCollector rather than a buffer, added test

Signed-off-by: Rohin Garg <rohin@character.ai>

rohingarg-c commented:
I updated the PR to reflect some of the comments. In particular

- removed the feature from the v0 engine
- added an additional test to ensure that the parameter is working correctly
- removed the extra buffer and instead modified the RequestOutputCollector to handle the stream_n logic

rohingarg-c committed:
minor bug fixes

Signed-off-by: Rohin Garg <rohin@character.ai>

rohingarg-c committed:
Merge branch 'main' of https://github.com/vllm-project/vllm into upstream-add-stream-n

rohingarg-c committed:
minor bug fix

Signed-off-by: Rohin Garg <rohin@character.ai>

rohingarg-c committed:
more bug fixes

Signed-off-by: Rohin Garg <rohin@character.ai>

rohingarg-c commented:
I’ve fixed all the build errors, and everything should now be ready to merge. Would appreciate a review!

njhill reviewed (commented):
Thanks @rohingarg-c and sorry for the delay.

rohingarg-c committed:
made stream_n engine args, fixed minor formatting issues

Signed-off-by: Rohin Garg <rohin@character.ai>

rohingarg-c committed:
Merge branch 'main' of https://github.com/vllm-project/vllm into upstream-add-stream-n

rohingarg-c committed:
fixed precommit issues

Signed-off-by: Rohin Garg <rohin@character.ai>

rohingarg-c committed:
Merge branch 'main' of https://github.com/vllm-project/vllm into upstream-add-stream-n

rohingarg-c committed:
rebased

Signed-off-by: Rohin Garg <rohin@character.ai>
